### PR TITLE
Wcsplotter/PVimage

### PR DESCRIPTION
**Pull request by CNOCycle**
This pull request is based on #40.

Displaying label in ImageViewer is improved and CARTA can show PV image.

Known issue:
1. Carta dose not have a class for spectral system(VRAD,VOPT,FREQ, ... ...)
2. Carta dose not have a class for spectral unit(m/s, km/s, Hz, GHz ... ...)
3. casacore always assumes the spectral axis is freq(Hz).
4. The spectral system is set 'system=VRAD,unit=km/s'(Known issue NO.1, NO.2) but it can be modified easily.
5. In UI side(`carta/cpp/core/Data/Image/Grid/*`), it assumes the image is celestial system and first axis is RA and second axis is DEC. It should be rewritten.

I will try to fix those known issues in the future.

### CODE REVIEW

**Comment by grimmerk** [comment]
Here is the change list. 
[Merge commit 'ecchen/fixedLoadCasaImage' into WCSPlotter/PVimage](https://github.com/CARTAvis/carta/pull/42/commits/528e950e43c10e31c35a20a9c52984f4e4da51e3)
[rewrite DisplayLabel in WCSplot plugin](https://github.com/CARTAvis/carta/pull/42/commits/03117528ad16732404e9c4f9f023a40033456ada)
[add a comment](https://github.com/CARTAvis/carta/pull/42/commits/778fe7a8fb5460fcef1427797243cfcc4e03d237)
[enhance displaying Label in WCSPlotter](https://github.com/CARTAvis/carta/pull/42/commits/4de77f7b87cf7a9f9ceb46f61e1c06f946a81f78)
[Display Label(FK5/FK4 to J2000/B1950)](https://github.com/CARTAvis/carta/pull/42/commits/4fc3c230ed0f2be75302b4a3f178a841d69388bf)
[improvements for displaying label in WCSplotter plugin](https://github.com/CARTAvis/carta/pull/42/commits/902707e4722d7220c9401e08b6e33eefdd906102)

Here are the changes which are not included in https://github.com/CARTAvis/carta/pull/43. PVimage/cursor
[remove 'm_axisDisplayInfos' in AstGridPlotter class](https://github.com/CARTAvis/carta/pull/42/commits/bedbbf7cd229f2da0df2ece33ebd4aea5d2c321d)
[remove _make2dFrameCelestial in AstGridPlotter.h](https://github.com/CARTAvis/carta/pull/42/commits/d376c67aaac771278a8a8a4bcd269e9561b50dab)
[modify comment in FitsHeaderExtractor](https://github.com/CARTAvis/carta/pull/42/commits/3af91865612f1212ed58c5806576bca19f3c9ec8)
[fix for warning [-Wformat-security] in WCSPlotter](https://github.com/CARTAvis/carta/pull/42/commits/cd2cf56cb51d9bed5fd0866097894af004cdcf1b)
[fix for warning [-Wformat-zero-length] in WCSPlotter](https://github.com/CARTAvis/carta/pull/42/commits/7bac04cd23df10077d0d958234ee077357607c0a)

File: carta/cpp/plugins/WcsPlotter/AstGridPlotter.cpp [hunk 1]
```diff
@@ -331,10 +331,7 @@ AstGridPlotter::plot()
     AstGuard astGuard;
 
     // ask AST to read in the FITS header
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wformat-zero-length"
-    AstFitsChan * fitschan = astFitsChan( NULL, NULL, "" );
-#pragma GCC diagnostic pop
+    AstFitsChan * fitschan = astFitsChan( NULL, NULL, "%s", "" );
```
Comment (grimmerk) [inline carta/cpp/plugins/WcsPlotter/AstGridPlotter.cpp]: what is the reason the will eliminate the warning?

**Comment by grimmerk** [comment]
add a comment -> The description is not clear....

File: carta/cpp/core/Data/Image/LayerData.cpp [hunk 2]
```diff
@@ -797,13 +797,8 @@ void LayerData::_renderStart(){
...
 
-    //Which display axes will be drawn.
-    AxisInfo::KnownType xType = m_dataSource->_getAxisXType();
-    AxisInfo::KnownType yType = m_dataSource->_getAxisYType();
-    QString displayLabelX = AxisMapper::getPurpose( xType, cs );
-    QString displayLabelY = AxisMapper::getPurpose( yType, cs );
-    gridService->setAxisLabel( 0, displayLabelX );
-    gridService->setAxisLabel( 1, displayLabelY );
+    // Label is set automatically and it is from the CTYPE which is defined in FITS header
+    // We need to set the spactral systm and unit
```
Comment (grimmerk) [inline carta/cpp/core/Data/Image/LayerData.cpp]: typo. systm -> system

File: carta/cpp/core/Data/Image/LayerData.cpp [hunk 3]
```diff
@@ -797,13 +797,8 @@ void LayerData::_renderStart(){
...
     }
 
-    //Which display axes will be drawn.
-    AxisInfo::KnownType xType = m_dataSource->_getAxisXType();
-    AxisInfo::KnownType yType = m_dataSource->_getAxisYType();
-    QString displayLabelX = AxisMapper::getPurpose( xType, cs );
-    QString displayLabelY = AxisMapper::getPurpose( yType, cs );
-    gridService->setAxisLabel( 0, displayLabelX );
-    gridService->setAxisLabel( 1, displayLabelY );
+    // Label is set automatically and it is from the CTYPE which is defined in FITS header
```
Comment (grimmerk) [inline carta/cpp/core/Data/Image/LayerData.cpp:800]: Which code is handling this?

File: carta/cpp/plugins/WcsPlotter/AstGridPlotter.cpp [hunk 4]
```diff
@@ -438,6 +438,7 @@ AstGridPlotter::plot()
         astClearStatus;
     }*/
 
+    // TODO: This part of code will move to '_setDisplayLabelOptionforAst()' in AstWcsGridRendeerService.cpp
```
Comment (grimmer0125) [inline carta/cpp/plugins/WcsPlotter/AstGridPlotter.cpp]: Is it done?

Comment (grimmerk) [inline carta/cpp/plugins/WcsPlotter/AstGridPlotter.cpp]: Is it done?

**Review by grimmerk — COMMENTED** [review]
@CNOCycle , this commit needs your introduction.

**Review by grimmerk — COMMENTED** [review]
@CNOCycle , this commit needs your introduction. Also sgp->setPlotOption("MinTickLen(1)=0");?

**Comment by grimmerk** [comment]
@kswang1029 
Does this pull request solves
1. show offset (arcsec)
2. frequency -> velocity (km/s)?

**Comment by kswang1029** [comment]
no, I think it is PVimage/UILabel which show correctly the xylabels as offset and velocity. @CNOCycle could you confirm?

**Comment by grimmerk** [comment]
PVimage/UILabel only fixes the label in ImageSetting.

**Comment by kswang1029** [comment]
oh you are right, this PR fixes 
1) show offset
2) frequency -> velocity

**Comment by grimmerk** [comment]
@CNOCycle: 
**rewrite DisplayLabel in WCSplot plugin** enhances:
1. add more check, e.g. //the line: e.g. check system for spectral
2. more organized codes (~refactor)

File: carta/cpp/core/Data/Image/LayerData.cpp (hunk 3 above)
Comment (grimmerk) [reply carta/cpp/core/Data/Image/LayerData.cpp:800]: @CNOCycle :
automaticlaly: setFitsHeader(_getFitsHeaderForAst)m)) )  + a little modification on ast, e.g. 
setPlotOption << QString("system(%1)=%2").arg(ii+1).arg( "VRAD" );
But we may totoal use casacore to get the whole info to feed to ast

File: carta/cpp/plugins/WcsPlotter/AstGridPlotter.cpp (hunk 4 above)
Comment (grimmerk) [reply carta/cpp/plugins/WcsPlotter/AstGridPlotter.cpp]: not done.

**Comment by grimmerk** [comment]
@kswang1029 , @appolloford 

After I select `aj.fits` to read, the grid line becomes **red**. After confirming, #40 does not have this problem. Should solve this.

**Comment by kswang1029** [comment]
With origin/mergeWCSPlotter branch in docker, I don't see this issue. I launch in server mode. Then open aJ.fits. I see the attached image. (I also applied clip just to make sure I do see the image)
screen shot 2017-04-20 at 12 12 09 pm

**Comment by grimmerk** [comment]
I think possible reasons:
1. different ast version?
2. desktop vs server?  
3. mac vs linux? 
And @markccchiang may help to debug this since he has Mac and Linux at the same time, using the same build way.

**Comment by CNOCycle** [comment]
The gridline is red because AST can not set some parameter successfully and AST uses default options which are defined in WCSPlotter plugin. 
I think that the error message is printed on the log.
Does CARTA complain about something?

**Comment by markccchiang** [comment]
Thanks for the comment! Following is the complain message from CARTA:

"!! AST: Error at line 405 in file
/Users/mark/cartawork/CARTAvis/carta/cpp/plugins/WcsPlotter/AstGridPlotter.cpp."
"!  astSet: The attribute setting "system=FK5, Equinox=J2000" is invalid
for the given CmpFrame."

I changed the line 405 in "AstGridPlotter.cpp" for "astSet(
plot, %s, stdstr.c_str() );" back to "astSet( plot, stdstr.c_str() );" and
the grid lines become white.

2017-04-20 12:47 GMT+08:00 CNOCycle <notifications@github.com>:

**Comment by kswang1029** [comment]
Could you compare the details of the gridline box? Please attach plots which show red box and normal white box based on your code-reverse tweak. Specifying "system" and "equinox" are critical to plot the gridline correctly.
 "!  astSet: The attribute setting "system=FK5, Equinox=J2000" is invalid for the given CmpFrame."

**Comment by markccchiang** [comment]
Did you mean these two plots?

white_box
red_box

**Comment by kswang1029** [comment]
No, I thought Grimmer saw the whole gridline plus the coordinate box are in red. Am I wrong?

**Comment by markccchiang** [comment]
The whole gridline plus the coordinate box are in red is what I have seen before modifying the code "AstGridPlotter.cpp", like the following plot

original_red_box

**Pull request merged**